topdown/json.go

```diff
@@ -230,7 +233,335 @@ func pathsToObject(paths []ast.Ref) ast.Object {
+	value, err := jsonPatchGet(target, from)
+	if err != nil {
+		return nil, err
+	} else if value == nil {
+		return nil, nil
+	}
+
+	return jsonPatchAdd(target, path, value)
+}
+
+func jsonPatchTest(target *ast.Term, path ast.Ref, value *ast.Term) (*ast.Term, error) {
+	actual, err := jsonPatchGet(target, path)
+	if err != nil {
+		return nil, err
+	}
+
+	if actual != nil && actual.Equal(value) {
+		return target, nil
+	}
+
+	return nil, nil
+}
+
+func builtinJSONPatch(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// JSON patch supports arrays, objects as well as values as the target.
+	target := ast.NewTerm(operands[0].Value)
+
+	// Expect an array of operations.
+	operations, err := builtins.ArrayOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	// Apply operations one by one.
+	for i := 0; i < operations.Len(); i++ {
+		if object, ok := operations.Elem(i).Value.(ast.Object); ok {
+			// Parse operation.
+			opTerm := object.Get(ast.StringTerm("op"))
+			if opTerm == nil {
+				return builtins.NewOperandErr(2, "must be an array of JSON-Patch objects")
+			}
+			op, ok := opTerm.Value.(ast.String)
+			if !ok {
+				return builtins.NewOperandErr(2, "must be an array of JSON-Patch objects")
+			}
+
+			// Parse path.
+			pathTerm := object.Get(ast.StringTerm("path"))
+			if pathTerm == nil {
+				return builtins.NewOperandErr(2, "patch is missing 'path' attribute")
+			}
+			path, err := parsePath(pathTerm)
+			if err != nil {
+				return err
+			}
+
+			value := object.Get(ast.StringTerm("value"))
+			switch op {
+			case "add":
+				if value == nil {
+					return builtins.NewOperandErr(2, "patch is missing 'value' attribute")
+				}
+				if target, err = jsonPatchAdd(target, path, value); err != nil {
+					return err
+				}
+			case "remove":
+				if target, _, err = jsonPatchRemove(target, path); err != nil {
+					return err
+				}
+			case "replace":
+				if value == nil {
+					return builtins.NewOperandErr(2, "patch is missing 'value' attribute")
+				}
+				if target, err = jsonPatchReplace(target, path, value); err != nil {
+					return err
+				}
+			case "move":
+				fromTerm := object.Get(ast.StringTerm("from"))
+				if fromTerm == nil {
+					return builtins.NewOperandErr(2, "patch is missing 'from' attribute")
+				}
+				from, err := parsePath(fromTerm)
+				if err != nil {
+					return err
+				}
+				if target, err = jsonPatchMove(target, path, from); err != nil {
+					return err
+				}
+			case "copy":
+				fromTerm := object.Get(ast.StringTerm("from"))
+				if fromTerm == nil {
+					return builtins.NewOperandErr(2, "patch is missing 'from' attribute")
+				}
+				from, err := parsePath(fromTerm)
+				if err != nil {
+					return err
+				}
+				if target, err = jsonPatchCopy(target, path, from); err != nil {
+					return err
+				}
+			case "test":
+				if target, err = jsonPatchTest(target, path, value); err != nil {
```

Comment:
❓ No `builtins.NewOperandErr(2, "patch is missing 'value' attribute")` for a missing value here?

Reply:
Fixed!